### PR TITLE
[ROS2] grid_map_pcl

### PR DESCRIPTION
**Pull request by Marwan99**
Main changes:
- Changing the launch file to a python launch script.
- Installing extra files in the package share directory so that they can be found using `ament_index_cpp::get_package_share_directory` which is the function I found as replacement for `ros::package::getPath`. @SteveMacenski please let me know if there is a better option.

@maximilianwulf  Want to know your thoughts on providing a pcd file in the package and use it to run the launch file as a demo, as currently the launch file just does nothing when it runs, but maybe I am missing something.

### CODE REVIEW

File: grid_map_pcl/CMakeLists.txt [hunk 1]
```diff
@@ -100,38 +84,78 @@ install(
...
+install(
+  DIRECTORY
+    config
+    launch
+  DESTINATION share/${PROJECT_NAME}
+)
+
+install(
+  DIRECTORY test/test_data/
+  DESTINATION share/${PROJECT_NAME}/test_data/
```
Comment (SteveMacenski) [inline grid_map_pcl/CMakeLists.txt:108]: Do we really want to install test data in the binaries?

**Review by SteveMacenski — APPROVED** [review]
LGTM

File: grid_map_pcl/CMakeLists.txt (hunk 1 above)
Comment (Marwan99) [reply grid_map_pcl/CMakeLists.txt:108]: Currently using `ament_index_cpp::get_package_share_directory` according to this [answer](https://answers.ros.org/question/288501/ros2-equivalent-of-rospackagegetpath/) to locate the test data, that's why test_data is being installed. Is there a different way?

Comment (SteveMacenski) [reply grid_map_pcl/CMakeLists.txt:108]: and that's needed to run the test in CI? Maybe put this install macro in the `if testing`?

**Comment by Marwan99** [comment]
I am struggling with getting [GridMapPclLoaderTest.NoisyStepTerrain](https://github.com/Marwan99/grid_map/blob/84a8e6baf5db3dffaa5c6a97829b73286506f549/grid_map_pcl/test/GridMapPclLoaderTest.cpp#L176-L205) to pass, it just crashes atm as in [here](https://app.circleci.com/pipelines/github/ANYbotics/grid_map/139/workflows/560c0648-85c4-4300-9631-7e80b723d524/jobs/146).

My observations/debugging attempts:
- When [GridMapPclLoaderTest.NoisyPlane](https://github.com/Marwan99/grid_map/blob/84a8e6baf5db3dffaa5c6a97829b73286506f549/grid_map_pcl/test/GridMapPclLoaderTest.cpp#L96-L119) is commented the test doesn't crash and passes.
- The test consistently fails on the 7th/8th iteration of the loop, despite the loop content being independent from the loop iteration number as far as I can tell.
- 98% of the time the code crashes [here](https://github.com/Marwan99/grid_map/blob/84a8e6baf5db3dffaa5c6a97829b73286506f549/grid_map_pcl/src/PointcloudProcessor.cpp#L53), I verified that the inputCloud is not empty when the test crashes. 2-3 of times (out of 10 of runs) it got past this line and crashed elsewhere.
- I dumped the contents of the test into a node (without the `EXPECT` statements) and the code ran just fine.

No really sure what to try next, any suggestions?

File: grid_map_pcl/CMakeLists.txt (hunk 1 above)
Comment (Marwan99) [reply grid_map_pcl/CMakeLists.txt]: Yes, needed to run tests. Installed in `if testing`.

**Comment by SteveMacenski** [comment]
I think this is a timeout issue, the timestamps are almost exactly 60s in length

**Comment by Marwan99** [comment]
Sorry, not really sure what you mean by "the timestamps are almost exactly 60s in length", do you mean the difference between each iteration (which looks like 6s to me)?

> I am struggling with getting GridMapPclLoaderTest.NoisyStepTerrain to pass, it just crashes atm as in here.

**Correction**, the test [here](https://app.circleci.com/pipelines/github/ANYbotics/grid_map/139/workflows/560c0648-85c4-4300-9631-7e80b723d524/jobs/146) is crashing at `GridMapPclLoaderTest.NoisyPlane` not `GridMapPclLoaderTest.NoisyStepTerrain`

**Comment by SteveMacenski** [comment]
If you look at the `INFO` timestamps, take the first one and the last one and subtract them. What you find is they're almost exactly 60s (`59.606`), minus setup time for the test, the likely cause of this failure is that the test is timing out due to a default 60s timeout. You can extend the timeout period for the test to give it more time to process.

**Comment by SteveMacenski** [comment]
Well now it passes, but not deep code😀

**Comment by Marwan99** [comment]
Retry DeepCode

**Comment by Marwan99** [comment]
Thanks for your help @SteveMacenski!

Feels like that the code checks are conscious and they make sure that at least one of them fail :')
Anyway I messaged deepcode about this.

**Comment by Arvi3d** [comment]
Deepcode retry

**Comment by maximilianwulf** [comment]
I didn't know that Deepcode is based in Zurich, I might drop by and ask what is going on :D

**Comment by Arvi3d** [comment]
I'm sorry for this inconvenience. Thanks to recent news https://techcrunch.com/2020/09/24/snyk-acquires-deepcode-to-boost-its-code-review-smarts/ we have a spike in traffic.

Working now to solve it.

File: grid_map_pcl/include/grid_map_pcl/PclLoaderParameters.hpp [hunk 2]
```diff
@@ -6,92 +6,108 @@
...
+    bool isDownsampleCloud_ = false;
+  };
+
+  struct ClusterExtractionParameters
+  {
+    double clusterTolerance_ = 0.3;
+    unsigned int minNumPoints_ = 2;
+    unsigned int maxNumPoints_ = 1000000;
+  };
+  struct OutlierRemovalParameters
```
Comment (maximilianwulf) [inline grid_map_pcl/include/grid_map_pcl/PclLoaderParameters.hpp:45]:
```suggestion

  struct OutlierRemovalParameters
```

File: grid_map_pcl/include/grid_map_pcl/PclLoaderParameters.hpp [hunk 3]
```diff
@@ -6,92 +6,108 @@
...
+  {
+    bool isRemoveOutliers_ = false;
+    double meanK_ = 10.0;
+    double stddevThreshold_ = 1.0;
+  };
+  struct RigidBodyTransformation
```
Comment (maximilianwulf) [inline grid_map_pcl/include/grid_map_pcl/PclLoaderParameters.hpp:51]:
```suggestion

  struct RigidBodyTransformation
```

File: grid_map_pcl/src/GridMapPclLoader.cpp [hunk 4]
```diff
@@ -99,33 +105,39 @@ void GridMapPclLoader::initializeGridMapGeometryFromInputCloud()
...
+    (maxBound.y + minBound.y) / 2.0);
   workingGridMap_.setGeometry(length, resolution, position);
 
-  ROS_INFO_STREAM(
-      "Grid map dimensions: " << workingGridMap_.getLength()(0) << " x " << workingGridMap_.getLength()(1));
-  ROS_INFO_STREAM("Grid map resolution: " << workingGridMap_.getResolution());
-  ROS_INFO_STREAM(
-      "Grid map num cells: " << workingGridMap_.getSize()(0) << " x " << workingGridMap_.getSize()(1));
-  ROS_INFO_STREAM("Initialized map geometry");
+  RCLCPP_INFO_STREAM(
```
Comment (maximilianwulf) [inline grid_map_pcl/src/GridMapPclLoader.cpp:117]: Nothing to change here, but it gets quite ugly to read.

File: grid_map_pcl/src/GridMapPclLoader.cpp [hunk 5]
```diff
@@ -99,33 +105,39 @@ void GridMapPclLoader::initializeGridMapGeometryFromInputCloud()
...
   // Preprocess: allocate memory in the internal data structure
   preprocessGridMapCells();
   workingGridMap_.add(layer);
-  grid_map::Matrix& gridMapData = workingGridMap_.get(layer);
+  grid_map::Matrix & gridMapData = workingGridMap_.get(layer);
   unsigned int linearGridMapSize = workingGridMap_.getSize().prod();
 
 #ifndef GRID_MAP_PCL_OPENMP_FOUND
-  ROS_WARN_STREAM("OpemMP not found, defaulting to single threaded implementation");
+  RCLCPP_WARN_STREAM("OpemMP not found, defaulting to single threaded implementation");
```
Comment (maximilianwulf) [inline grid_map_pcl/src/GridMapPclLoader.cpp:140]: Why not add `node_logger_` here?

File: grid_map_pcl/src/PclLoaderParameters.cpp [hunk 6]
```diff
@@ -8,91 +8,98 @@
...
-	parameters_.downsampling_.voxelSize_.y() =
-			yamlNode[prefix]["downsampling"]["voxel_size"]["y"].as<double>();
-	parameters_.downsampling_.voxelSize_.z() =
-			yamlNode[prefix]["downsampling"]["voxel_size"]["z"].as<double>();
-
+#include <rclcpp/rclcpp.hpp>
+
+#include <string>
+
+
```
Comment (maximilianwulf) [inline grid_map_pcl/src/PclLoaderParameters.cpp:15]:
```suggestion
```

File: grid_map_pcl/src/grid_map_pcl_loader_node.cpp [hunk 7]
```diff
@@ -1,12 +1,17 @@
 /*
- * grid_map_pcl_loader_node.cpp
+ * grid_map_pcl_loader_node->cpp
```
Comment (maximilianwulf) [inline grid_map_pcl/src/grid_map_pcl_loader_node.cpp:2]: ?

File: grid_map_pcl/src/grid_map_pcl_loader_node.cpp [hunk 8]
```diff
@@ -15,34 +20,32 @@
...
+int main(int argc, char ** argv)
+{
+  rclcpp::init(argc, argv);
+  auto node = std::make_shared<rclcpp::Node>("grid_map_pcl_loader_node");
+  gm::setVerbosityLevelToDebugIfFlagSet(node);
 
-  ros::Publisher gridMapPub;
-  gridMapPub = nh.advertise<grid_map_msgs::GridMap>("grid_map_from_raw_pointcloud", 1, true);
+  auto gridMapPub = node->create_publisher<grid_map_msgs::msg::GridMap>(
+    "grid_map_from_raw_pointcloud", rclcpp::QoS(1).transient_local());
```
Comment (maximilianwulf) [inline grid_map_pcl/src/grid_map_pcl_loader_node.cpp:30]: No action required, `rclcpp::QoS(1).transient_local()` I think in terms of readability this is a downgrade

File: grid_map_pcl/src/grid_map_pcl_loader_node.cpp [hunk 9]
```diff
@@ -15,34 +20,32 @@
...
 namespace gm = ::grid_map::grid_map_pcl;
 
-int main(int argc, char** argv) {
-  ros::init(argc, argv, "grid_map_pcl_loader_node");
-  ros::NodeHandle nh("~");
-  gm::setVerbosityLevelToDebugIfFlagSet(nh);
+int main(int argc, char ** argv)
+{
+  rclcpp::init(argc, argv);
+  auto node = std::make_shared<rclcpp::Node>("grid_map_pcl_loader_node");
```
Comment (maximilianwulf) [inline grid_map_pcl/src/grid_map_pcl_loader_node.cpp:26]: I like the practice of `nh` -> `node` much cleaner.

File: grid_map_pcl/src/grid_map_pcl_loader_node.cpp [hunk 10]
```diff
@@ -15,34 +20,32 @@
...
+  gridMap.setFrameId(gm::getMapFrame(node));
 
-  grid_map_msgs::GridMap msg;
-  grid_map::GridMapRosConverter::toMessage(gridMap, msg);
-  gridMapPub.publish(msg);
+  gm::saveGridMap(gridMap, node, gm::getMapRosbagTopic(node));
 
+  // publish grid map
+  auto msg = grid_map::GridMapRosConverter::toMessage(gridMap);
+  gridMapPub->publish(std::move(msg));
```
Comment (maximilianwulf) [inline grid_map_pcl/src/grid_map_pcl_loader_node.cpp:46]: Are you sure the `std::move` is doing the intended thing here? 

which `publish()` function would it use in this case? https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/publisher.hpp#L185

File: grid_map_pcl/test/GridMapPclLoaderTest.cpp [hunk 11]
```diff
@@ -6,89 +6,110 @@
...
 {
-  grid_map_pcl_test::setVerbosityLevel(ros::console::levels::Warn);
+class GridMapPclLoaderTest : public ::testing::Test
+{
+protected:
+  void SetUp() override
+  {
+    node = std::make_shared<rclcpp::Node>("GridMapPclLoaderTest");
+
+    // Changing the logging verbosity to WARN
```
Comment (maximilianwulf) [inline grid_map_pcl/test/GridMapPclLoaderTest.cpp:32]: I think you mean `DEBUG` ?

File: grid_map_pcl/test/test_helpers.cpp [hunk 12]
```diff
@@ -75,20 +80,19 @@ Pointcloud::Ptr createNormallyDistributedBlobOfPoints(unsigned int nPoints, doub
...
 {
-
   const double upperBound = maxXY;
   const double lowerBound = minXY;
   std::uniform_real_distribution<double> uniformDist(lowerBound, upperBound);
   const double mean = meanZ;
   const double stdDev = stdDevZ;
   std::normal_distribution<double> normalDist(mean, stdDev);  // N
-  //std::cout << distr(generator) << '\n';
+  // std::cout << distr(generator) << '\n';
```
Comment (maximilianwulf) [inline grid_map_pcl/test/test_helpers.cpp:95]:
```suggestion
```

File: grid_map_pcl/test/test_helpers.hpp [hunk 13]
```diff
@@ -6,56 +6,66 @@
...
-#include "grid_map_core/GridMap.hpp"
-#include "grid_map_core/iterators/GridMapIterator.hpp"
 #include <pcl/common/common.h>
-#include <ros/package.h>
-#include <ros/console.h>
+#include <ament_index_cpp/get_package_share_directory.hpp>
+#include <string>
+#include <vector>
 #include <random>
+#include "grid_map_core/GridMap.hpp"
```
Comment (maximilianwulf) [inline grid_map_pcl/test/test_helpers.hpp:18]:
```suggestion

#include "grid_map_core/GridMap.hpp"
```

File: grid_map_pcl/test/HelpersTest.cpp [hunk 14]
```diff
@@ -6,28 +6,30 @@
...
+namespace grid_map
+{
+namespace grid_map_pcl_test
+{
 
 TEST(HelpersTest, MeanPositionTest)
 {
-  grid_map_pcl_test::setVerbosityLevel(ros::console::levels::Warn);
-  const auto seed = rand();
+  // rcutils_logging_set_logger_level(get_logger().get_name(), RCUTILS_LOG_SEVERITY_WARN);
```
Comment (maximilianwulf) [inline grid_map_pcl/test/HelpersTest.cpp:27]: Delete?

File: grid_map_pcl/test/GridMapPclLoaderTest.cpp [hunk 15]
```diff
@@ -6,89 +6,110 @@
...
+    if (ret != RCUTILS_RET_OK) {
+      RCLCPP_ERROR(
+        node->get_logger(), "Failed to change logging severity: %s",
+        rcutils_get_error_string().str);
+      rcutils_reset_error();
+    }
+  }
 
-  grid_map::GridMapPclLoader gridMapPclLoader;
+  rclcpp::Node::SharedPtr node;
```
Comment (maximilianwulf) [inline grid_map_pcl/test/GridMapPclLoaderTest.cpp:43]:
```suggestion
  rclcpp::Node::SharedPtr node_;
```

**Review by maximilianwulf — COMMENTED** [review]
Thanks a lot for the MR, as usual only minor stuff.

**Comment by Arvi3d** [comment]
Deepcode retry

**Comment by maximilianwulf** [comment]
@Arvi3d thanks, what was the problem?

**Comment by Arvi3d** [comment]
> @Arvi3d thanks, what was the problem?

Due to high load, our analyzer could not get results within specified time limit. We fixed bottleneck and upscaled our servers.

File: grid_map_pcl/test/GridMapPclLoaderTest.cpp (hunk 15 above)
Comment (Marwan99) [reply grid_map_pcl/test/GridMapPclLoaderTest.cpp]: I noticed that the logger of the node was the only thing used from the node so I swapped the node with a logger instead. Made sure to include an "_" at the end of the variable name.

File: grid_map_pcl/test/GridMapPclLoaderTest.cpp (hunk 11 above)
Comment (Marwan99) [reply grid_map_pcl/test/GridMapPclLoaderTest.cpp:30]: I think originally the logging level was being changed to `WARN`, so I think the comment is correct, but I messed up the actual change of the logging level so fixed it now to`WARN`. Does that sound right?

File: grid_map_pcl/src/grid_map_pcl_loader_node.cpp (hunk 8 above)
Comment (Marwan99) [reply grid_map_pcl/src/grid_map_pcl_loader_node.cpp]: Moved the setting of the QoS to a variable, hope that helps a bit.

File: grid_map_pcl/src/grid_map_pcl_loader_node.cpp (hunk 7 above)
Comment (Marwan99) [reply grid_map_pcl/src/grid_map_pcl_loader_node.cpp]: The pitfall of using replace all. Fixed :)

File: grid_map_pcl/include/grid_map_pcl/PclLoaderParameters.hpp (hunk 2 above)
Comment (Marwan99) [reply grid_map_pcl/include/grid_map_pcl/PclLoaderParameters.hpp:46]: That should be fixed, not sure why not shown as outdated.

File: grid_map_pcl/include/grid_map_pcl/PclLoaderParameters.hpp (hunk 3 above)
Comment (Marwan99) [reply grid_map_pcl/include/grid_map_pcl/PclLoaderParameters.hpp:53]: That should be fixed, not sure why not shown as outdated.

File: grid_map_pcl/src/grid_map_pcl_loader_node.cpp (hunk 10 above)
Comment (Marwan99) [reply grid_map_pcl/src/grid_map_pcl_loader_node.cpp:47]: The one you highlighted is the one being used.
Without `std::move` the code doesn't compile. What do you think might be wrong?

**Comment by maximilianwulf** [comment]
> Due to high load, our analyzer could not get results within specified time limit. We fixed bottleneck and upscaled our servers.

Okay, thanks. Good to hear that you have many customers.

**Comment by Marwan99** [comment]
@maximilianwulf Please review the changes.

File: grid_map_pcl/src/grid_map_pcl_loader_node.cpp (hunk 10 above)
Comment (maximilianwulf) [reply grid_map_pcl/src/grid_map_pcl_loader_node.cpp:47]: All good, then.

File: grid_map_pcl/test/GridMapPclLoaderTest.cpp (hunk 11 above)
Comment (maximilianwulf) [reply grid_map_pcl/test/GridMapPclLoaderTest.cpp:30]: Yes.

**Review by maximilianwulf — APPROVED** [review]
Thanks. LGTM.

**Comment by Marwan99** [comment]
@SteveMacenski Good to go?

**Pull request merged**